src/jvm/main/org/jetbrains/kotlinx/lincheck/bootstrap/CodeInformation.kt

```diff
@@ -0,0 +1,140 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2024 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+// we need to use some "legal" package for the bootstrap class loader
+@file:Suppress("PackageDirectoryMismatch")
+
+package sun.nio.ch.lincheck
+
+import org.jetbrains.kotlinx.lincheck.TransformationClassLoader.REMAPPED_PACKAGE_INTERNAL_NAME
+import org.jetbrains.kotlinx.lincheck.transformation.canonicalClassName
+import sun.misc.*
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier.*
+import java.util.concurrent.atomic.*
+
+/**
+ * [CodeLocations] object is used to maintain the mapping between unique IDs and code locations.
+ * When Lincheck detects an error in the model checking mode, it provides a detailed interleaving trace.
+ * This trace includes a list of all shared memory events that occurred during the execution of the program,
+ * along with their corresponding code locations. To minimize overhead, Lincheck assigns unique IDs to all
+ * code locations it analyses, and stores more detailed information necessary for trace generation in this object.
+ */
+internal object CodeLocations {
```

Comment:
Would be nice to document why this (and other transformation-related classes) are implemented as singletons.

Reply:
It is a singleton because it's a convenient wrapper on a static content, removing it from the top-level.